dex/pair/src/safe_price_views.rs

```diff
@@ -0,0 +1,287 @@
+        } else {
+            let output_amount =
+                input_payment.amount * weighted_first_token_reserve / weighted_second_token_reserve;
+            EsdtTokenPayment::new(first_token_id, 0, output_amount)
+        };
+        require!(output_payment.amount > 0u64, ERROR_ZERO_AMOUNT);
+        output_payment
+    }
+
+    fn get_price_observation(
+        &self,
+        current_index: usize,
+        max_observations: usize,
+        price_observations: &ManagedVec<PriceObservation<Self::Api>>,
+        search_round: Round,
+    ) -> PriceObservation<Self::Api> {
+        if price_observations.is_empty() {
+            sc_panic!(ERROR_SAFE_PRICE_OBSERVATION_DOES_NOT_EXIST)
+        }
+
+        // Check if the requested price observation is the last one
+        let last_observation = price_observations.get(current_index);
+        if last_observation.recording_round <= search_round {
+            return last_observation;
+        }
+
+        // Check if the observation round exists in the list
+        let oldest_observation_index = if price_observations.len() < max_observations {
+            0
+        } else {
+            (current_index + 1) % max_observations
+        };
+        let oldest_observation_option = price_observations.try_get(oldest_observation_index);
+        match oldest_observation_option {
+            Some(oldest_observation) => {
+                if search_round == oldest_observation.recording_round {
+                    return oldest_observation;
+                }
+                require!(
+                    search_round > oldest_observation.recording_round,
+                    ERROR_SAFE_PRICE_OBSERVATION_DOES_NOT_EXIST
+                );
+            }
+            None => sc_panic!(ERROR_SAFE_PRICE_OBSERVATION_DOES_NOT_EXIST),
+        }
+
+        // Binary search algorithm
+        let mut search_index = 0;
+        let mut left_index;
+        let mut right_index;
+        let observation_at_index_0 = price_observations.get(0);
+        if observation_at_index_0.recording_round <= search_round {
+            left_index = 0;
+            right_index = current_index - 1;
+        } else {
+            left_index = current_index + 1;
+            right_index = max_observations - 1;
+        }
+
+        while left_index <= right_index {
+            search_index = (left_index + right_index) / 2;
+            let price_observation = price_observations.get(search_index);
+            match price_observation.recording_round.cmp(&search_round) {
+                Ordering::Equal => return price_observation,
+                Ordering::Less => left_index = search_index + 1,
+                Ordering::Greater => right_index = search_index - 1,
+            }
+        }
+
+        // Linear interpolation in case there is no price observation for the searched round
+        let last_found_observation = price_observations.get(search_index);
```

Comment:
move linear interpolation to a new function.

Reply:
Updated.